process/block/poolsCleaner/crossTxsPoolsCleaner.go

```diff
@@ -0,0 +1,306 @@
+	for {
+		time.Sleep(sleepTime)
+		numCrossTxsInMap := ctpc.cleanCrossTxsPoolsIfNeeded()
+		log.Debug("crossTxsPoolsCleaner.cleanCrossTxsPools", "num cross txs in map", numCrossTxsInMap)
+	}
+}
+
+func (ctpc *crossTxsPoolsCleaner) receivedBlockTx(key []byte) {
+	if key == nil {
+		return
+	}
+
+	log.Trace("crossTxsPoolsCleaner.receivedBlockTx", "hash", key)
+	ctpc.processReceivedTx(key, blockTx)
+}
+
+func (ctpc *crossTxsPoolsCleaner) receivedRewardTx(key []byte) {
+	if key == nil {
+		return
+	}
+
+	log.Trace("crossTxsPoolsCleaner.receivedRewardTx", "hash", key)
+	ctpc.processReceivedTx(key, rewardTx)
+}
+
+func (ctpc *crossTxsPoolsCleaner) receivedUnsignedTx(key []byte) {
+	if key == nil {
+		return
+	}
+
+	log.Trace("crossTxsPoolsCleaner.receivedUnsignedTx", "hash", key)
+	ctpc.processReceivedTx(key, unsignedTx)
+}
+
+func (ctpc *crossTxsPoolsCleaner) processReceivedTx(key []byte, txType int8) {
+	isCrossTx := false
+	var crossTxInfo *txInfo
+	for shardID := uint32(0); shardID < ctpc.shardCoordinator.NumberOfShards(); shardID++ {
+		isCrossTx, crossTxInfo = ctpc.isCrossTxFromShard(key, shardID, txType)
+		if isCrossTx {
+			break
+		}
+	}
+
+	if !isCrossTx {
+		isCrossTx, crossTxInfo = ctpc.isCrossTxFromShard(key, core.MetachainShardId, txType)
+		if !isCrossTx {
+			return
+		}
+	}
+
+	if crossTxInfo == nil {
+		return
+	}
+
+	ctpc.mutMapCrossTxsRounds.Lock()
+	defer ctpc.mutMapCrossTxsRounds.Unlock()
+
+	if _, ok := ctpc.mapCrossTxsRounds[string(key)]; !ok {
+		crossTxInfo.round = ctpc.rounder.Index()
+		crossTxInfo.txType = txType
+		ctpc.mapCrossTxsRounds[string(key)] = crossTxInfo
+
+		log.Trace("transaction has been added",
+			"hash", key,
+			"round", crossTxInfo.round,
+			"sender shard", crossTxInfo.senderShardID,
+			"receiver shard", crossTxInfo.receiverShardID,
+			"type", getTxTypeName(crossTxInfo.txType))
+	}
+}
+
+func (ctpc *crossTxsPoolsCleaner) isCrossTxFromShard(txHash []byte, shardID uint32, txType int8) (bool, *txInfo) {
+	transactionPool := ctpc.getTransactionPool(txType)
+	if transactionPool == nil {
+		return false, nil
+	}
+
+	selfShardID := ctpc.shardCoordinator.SelfId()
+	if shardID == selfShardID {
+		return false, nil
+	}
+
+	strCache := process.ShardCacherIdentifier(shardID, selfShardID)
+	txStore := transactionPool.ShardDataStore(strCache)
+	if txStore == nil {
+		return false, nil
+	}
+
+	_, ok := txStore.Peek(txHash)
+	if !ok {
+		return false, nil
+	}
+
+	return true, &txInfo{txStore: txStore, senderShardID: shardID, receiverShardID: selfShardID}
+}
+
+func (ctpc *crossTxsPoolsCleaner) cleanCrossTxsPoolsIfNeeded() int {
+	ctpc.mutMapCrossTxsRounds.Lock()
+	defer ctpc.mutMapCrossTxsRounds.Unlock()
+
+	selfShardID := ctpc.shardCoordinator.SelfId()
+	numPendingMiniBlocks := ctpc.blockTracker.GetNumPendingMiniBlocks(selfShardID)
+	numTxsCleaned := 0
+
+	for hash, crossTxInfo := range ctpc.mapCrossTxsRounds {
+		value, ok := crossTxInfo.txStore.Get([]byte(hash))
```

Comment:
do we even need to get the tx from txstore and verify that it is valid etc?
don't we just need to remove if enough time passed?
I would just check the time in the crossTxInfo and if it is time to remove do the remove and don't care
no need to access multiple times the store if it is not needed.

Reply:
It was done this way only to get the txs and to print some info about these which are going to be removed, but we can go without these info and to simplify this part.